File: sdks/flutter/ios/Classes/LocalNotificationBridge.swift
import Flutter
import Foundation
import Notifie
import UserNotifications

/**
 Translates method-channel arguments into the native SDK's API.

 Deliberately a translator and nothing more. Trigger construction, the pending
 limit and identifier namespacing all live in the Notifie SDK this delegates to;
 reimplementing any of it here would create a second engine to diverge from the
 first, and the divergence would be silent.
 */
enum LocalNotificationBridge {

    static func schedule(_ arguments: [String: Any]) async -> [String: Any] {
        let notification: LocalNotification
        do {
            notification = try decode(arguments)
        } catch let error as DecodeError {
            return ["error": "invalid_request", "message": error.reason]
        } catch {
            return ["error": "invalid_request", "message": String(describing: error)]
        }

        switch await Notifie.schedule(notification) {
        case .success:
            // iOS has no separately granted precision: a scheduled request is
            // as exact as the platform offers.
            return ["precision": "exact"]
        case .failure(let error):
            return failure(for: error)
        }
    }

    static func cancel(_ arguments: [String: Any]) {
        guard let id = arguments["id"] as? String else { return }
        Notifie.cancelScheduled(id: id)
    }

    static func pending() async -> [[String: Any]] {
        await Notifie.pendingScheduled().map { entry in
            var result: [String: Any] = ["id": entry.id]
            if let next = entry.nextTriggerDate {
                result["nextTrigger"] = Int(next.timeIntervalSince1970 * 1000)
            }
            return result
        }
    }

    static func capabilities() async -> [String: Any] {
        [
            "permission": await permissionState(),
            // iOS does not gate scheduling precision behind a permission the
            // way Android does, so this is always available.
            "canScheduleExactAlarms": true,
            "supportedSchedules": ["at", "after", "daily", "weekly"],
            // iOS keeps only the 64 soonest pending notifications.
            "pendingCapacity": LocalNotificationLimits.maxPending,
        ]
    }

    static func requestPermission() async -> String {
        let centre = UNUserNotificationCenter.current()
        do {
            _ = try await centre.requestAuthorization(options: [.alert, .badge, .sound])
        } catch {
            return "denied"
        }
        return await permissionState()
    }

    private static func permissionState() async -> String {
        switch await UNUserNotificationCenter.current().notificationSettings().authorizationStatus {
        case .authorized, .ephemeral: return "granted"
        case .provisional: return "provisional"
        case .denied: return "denied"
        default: return "notDetermined"
        }
    }

    private static func failure(for error: LocalScheduleError) -> [String: Any] {
        switch error {
        case .permissionDenied: return ["error": "permission_denied"]
        case .scheduleInPast: return ["error": "schedule_in_past"]
        case .capacityExceeded: return ["error": "capacity_exceeded"]
        case .invalidRequest(let reason):
            return ["error": "invalid_request", "message": reason]
        case .platformError(let reason):
            return ["error": "platform_error", "message": reason]
        }
    }

    // MARK: - Decoding

    struct DecodeError: Error {
        let reason: String
    }

    static func decode(_ arguments: [String: Any]) throws -> LocalNotification {
        guard let id = arguments["id"] as? String else {
            throw DecodeError(reason: "id is required")
        }
        guard let title = arguments["title"] as? String else {
            throw DecodeError(reason: "title is required")
        }
        guard let body = arguments["body"] as? String else {
            throw DecodeError(reason: "body is required")
        }
        guard let scheduleArguments = arguments["schedule"] as? [String: Any] else {
            throw DecodeError(reason: "schedule is required")
        }

        let customData = (arguments["customData"] as? [String: Any] ?? [:])
            .reduce(into: [String: String]()) { output, entry in
                output[entry.key] = String(describing: entry.value)
            }

        return LocalNotification(
            id: id,
            title: title,
            body: body,
            schedule: try decodeSchedule(scheduleArguments),
            deepLink: arguments["deepLink"] as? String,
            customData: customData,
            ios: decodeIosOptions(arguments["ios"] as? [String: Any])
        )
    }

    private static func decodeSchedule(_ arguments: [String: Any]) throws -> LocalSchedule {
        switch arguments["type"] as? String {
        case "at":
            guard let raw = arguments["timestamp"] as? String else {
                throw DecodeError(reason: "timestamp is required")
            }
            // Parsed with its explicit offset. A bare local timestamp would be
            // reinterpreted in the device's timezone.
            guard let date = iso8601.date(from: raw) ?? iso8601NoFraction.date(from: raw) else {
                throw DecodeError(reason: "timestamp must be ISO-8601 with an offset")
            }
            return .at(date)
        case "after":
            guard let seconds = arguments["seconds"] as? NSNumber else {
                throw DecodeError(reason: "seconds is required")
            }
            return .after(seconds: seconds.doubleValue)
        case "daily":
            return .daily(
                hour: try requireInt(arguments, "hour"),
                minute: try requireInt(arguments, "minute")
            )
        case "weekly":
            return .weekly(
                weekday: try requireInt(arguments, "weekday"),
                hour: try requireInt(arguments, "hour"),
                minute: try requireInt(arguments, "minute")
            )
        default:
            throw DecodeError(reason: "unknown schedule type")
        }
    }

    private static func decodeIosOptions(
        _ arguments: [String: Any]?
    ) -> LocalNotificationIOSOptions? {
        guard let arguments else { return nil }
        var sound: LocalNotificationSound?
        if let named = arguments["sound"] as? String {
            sound = .named(named)
        } else if let enabled = arguments["sound"] as? Bool {
            sound = enabled ? .default : .silent
        }
        return LocalNotificationIOSOptions(
            threadId: arguments["threadId"] as? String,
            categoryId: arguments["categoryId"] as? String,
            badge: (arguments["badge"] as? NSNumber)?.intValue,
            sound: sound
        )
    }

    private static func requireInt(_ arguments: [String: Any], _ key: String) throws -> Int {
        guard let value = arguments[key] as? NSNumber else {
            throw DecodeError(reason: "\(key) is required")
        }
        return value.intValue
    }

    private static let iso8601: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    private static let iso8601NoFraction = ISO8601DateFormatter()
}

File: sdks/flutter/android/src/main/java/dev/notifie/flutter/LocalNotificationBridge.kt
package dev.notifie.flutter

import android.content.Context
import dev.notifie.LocalNotification
import dev.notifie.LocalNotificationAndroidOptions
import dev.notifie.Notifie
import dev.notifie.LocalSchedule
import dev.notifie.LocalScheduleError
import dev.notifie.LocalSchedulePrecision
import dev.notifie.LocalScheduleResult

/**
 * Translates method-channel arguments into the native scheduler's API.
 *
 * Deliberately a translator and nothing more. Persistence, reboot recovery,
 * alarm precision and drift-free recurrence all live in the Android SDK this
 * delegates to; reimplementing any of it here would create a second engine to
 * diverge from the first, and the divergence would be silent.
 */
internal object LocalNotificationBridge {

    fun schedule(context: Context, arguments: Map<*, *>): Map<String, Any?> {
        val notification = try {
            decode(arguments)
        } catch (error: IllegalArgumentException) {
            return failure("invalid_request", error.message)
        }

        return when (val result = Notifie.schedule(context, notification)) {
            is LocalScheduleResult.Scheduled -> mapOf(
                // The granted precision, not the requested one: Android
                // downgrades exact alarms without the 12+ permission.
                "precision" to when (result.precision) {
                    LocalSchedulePrecision.EXACT -> "exact"
                    LocalSchedulePrecision.INEXACT -> "inexact"
                },
                "nextTrigger" to result.nextTriggerAtMillis,
            )
            is LocalScheduleResult.Failed -> failure(code(result.error), result.message)
        }
    }

    fun cancel(context: Context, arguments: Map<*, *>) {
        val id = arguments["id"] as? String ?: return
        Notifie.cancelScheduled(context, id)
    }

    fun pending(context: Context): List<Map<String, Any?>> =
        Notifie.pendingScheduled(context).map {
            mapOf("id" to it.id, "nextTrigger" to it.nextTriggerAtMillis)
        }

    fun capabilities(context: Context): Map<String, Any?> = mapOf(
        "permission" to permissionState(context),
        "canScheduleExactAlarms" to canScheduleExactAlarms(context),
        "supportedSchedules" to listOf("at", "after", "daily", "weekly"),
        // Android publishes no fixed pending-notification limit, unlike iOS.
        // Reporting null is honest; inventing a number would not be.
        "pendingCapacity" to null,
    )

    fun permissionState(context: Context): String {
        val manager = androidx.core.app.NotificationManagerCompat.from(context)
        return if (manager.areNotificationsEnabled()) "granted" else "denied"
    }

    private fun canScheduleExactAlarms(context: Context): Boolean {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.S) return true
        val manager = context.getSystemService(android.app.AlarmManager::class.java)
        return manager?.canScheduleExactAlarms() ?: false
    }

    private fun code(error: LocalScheduleError): String = when (error) {
        LocalScheduleError.INVALID_REQUEST -> "invalid_request"
        LocalScheduleError.PERMISSION_DENIED -> "permission_denied"
        LocalScheduleError.SCHEDULE_IN_PAST -> "schedule_in_past"
        LocalScheduleError.PLATFORM_ERROR -> "platform_error"
    }

    private fun failure(code: String, message: String?): Map<String, Any?> =
        mapOf("error" to code, "message" to message)

    internal fun decode(arguments: Map<*, *>): LocalNotification {
        val scheduleArgs = arguments["schedule"] as? Map<*, *>
            ?: throw IllegalArgumentException("schedule is required")

        val android = (arguments["android"] as? Map<*, *>).let { options ->
            LocalNotificationAndroidOptions(
                channelId = options?.get("channelId") as? String,
                exact = options?.get("exact") as? Boolean ?: false,
                allowWhileIdle = options?.get("allowWhileIdle") as? Boolean ?: false,
                groupKey = options?.get("groupKey") as? String,
            )
        }

        val data = (arguments["customData"] as? Map<*, *>).orEmpty()
            .entries
            .mapNotNull { (key, value) ->
                val name = key as? String ?: return@mapNotNull null
                name to value.toString()
            }
            .toMap()

        return LocalNotification(
            id = arguments["id"] as? String
                ?: throw IllegalArgumentException("id is required"),
            title = arguments["title"] as? String
                ?: throw IllegalArgumentException("title is required"),
            body = arguments["body"] as? String
                ?: throw IllegalArgumentException("body is required"),
            schedule = decodeSchedule(scheduleArgs),
            deepLink = arguments["deepLink"] as? String,
            customData = data,
            android = android,
        )
    }

    private fun decodeSchedule(arguments: Map<*, *>): LocalSchedule =
        when (val type = arguments["type"] as? String) {
            "at" -> {
                // Parsed as an instant with its explicit offset. A bare local
                // timestamp would be reinterpreted in the device's timezone.
                val raw = arguments["timestamp"] as? String
                    ?: throw IllegalArgumentException("timestamp is required")
                LocalSchedule.At(parseIso8601(raw))
            }
            "after" -> LocalSchedule.After(requireLong(arguments, "seconds"))
            "daily" -> LocalSchedule.Daily(
                hour = requireInt(arguments, "hour"),
                minute = requireInt(arguments, "minute"),
            )
            "weekly" -> LocalSchedule.Weekly(
                weekday = requireInt(arguments, "weekday"),
                hour = requireInt(arguments, "hour"),
                minute = requireInt(arguments, "minute"),
            )
            else -> throw IllegalArgumentException("unknown schedule type \"$type\"")
        }

    private fun parseIso8601(raw: String): Long = try {
        java.time.Instant.parse(raw).toEpochMilli()
    } catch (error: java.time.format.DateTimeParseException) {
        throw IllegalArgumentException("timestamp must be ISO-8601 with an offset")
    }

    private fun requireInt(arguments: Map<*, *>, key: String): Int =
        (arguments[key] as? Number)?.toInt()
            ?: throw IllegalArgumentException("$key is required")

    private fun requireLong(arguments: Map<*, *>, key: String): Long =
        (arguments[key] as? Number)?.toLong()
            ?: throw IllegalArgumentException("$key is required")
}
